caps/types.go

```diff
@@ -83,3 +86,52 @@ func (t *Type) MarshalJSON() ([]byte, error) {
 func (t *Type) UnmarshalJSON(data []byte) error {
 	return json.Unmarshal(data, &t.Name)
 }
+
+// GrantPermissions makes it possible for the package `snapName` to use the
+// resource described by the capability `cap`.
+func (t *Type) GrantPermissions(snapName string, cap *Capability) error {
+	// Ensure that the capability is valid
+	if err := t.Validate(cap); err != nil {
+		return err
+	}
+	// Grant all permissions required.
+	for i := range cap.Type.SecuritySystems {
+		sec := cap.Type.SecuritySystems[i]
+		if err := sec.GrantPermissions(snapName, cap); err != nil {
+			// If we already granted something, try to revoke that permission instead
+			// NOTE: "i - 1" because grant that fails should fail atomically
+			for j := i - 1; j >= 0; j-- {
+				sec := cap.Type.SecuritySystems[j]
+				if err := sec.RevokePermissions(snapName, cap); err != nil {
+					// XXX: Should we do something other than panic here?
+					panic(fmt.Sprintf("unable to revoke partially granted permissions: %q", err))
```

Comment:
eventually yes, there's got to be a big red "give up and work out the current state of the system from the top again" button in the Overseer